### ISSUE TITLE
Unity 2022.2+ DX11/12 3D Active Stereo

### ISSUE THREAD

**Issue by VRS3DGuru2** [Vital-Volkov/Stereoscopic-3D-system-for-Unity-2019-#7]
Wanted to move my comments over here and figure out ways to help contribute

**Comment by VRS3DGuru2** [comment]
I have looked at your plugin, its brilliant work. 

For Generic Active 3D Stereo Frame Sequential (Single Video Cable @120Hz+) in Unity 2022.2+ and DX11/12 - I could not figure out how to modify it to output 3D Active Stereo with DX11/12 on Unity 2022.2 Using Nvidia Geforce 2080 and newer GPUs ; although, I did try. I was also concerned about being able to consistently present the left eye frame to left eye of user each time an app starts. I hit a definite limit on my knowledge.

My project is using HDRP 14.0.6 with Unity 2022.2.13f1 preparing for the 2022.3 LTS. I plan on keeping to 2022.3 LTS for the project. I have verified that your approach works with all in-frame packed stereoscopic outputs you list, Interlacing, checkerboard, above below, side-by-side. I use DX11 for DLSS compatibility on Nvidia 2080+ GPUs. I use DX12 for Raytracing modes, which presently only runs with my content very well on a Nvidia 3080+ GPU. I need to optimize better for Raytracing. All of my selected volume over-rides for various effects work as expected with your approach to stereoscopic output. The Cinemachine cameras that I am using do not work with the default prefab you have, but I see there is a way to fix that.

**Comment by Vital-Volkov** [comment]
I'll add an independent sequential method for my plugin and we test it. As sequential is just switching the left/right camera image each frame so it is very easy to realise it in the engine.
But as I understand you trying to make it via Unity builtin S3D API and get a problem.

Only thing is how glassess know the current frame is left or right(I guess some flag available in NVidia driver) but I think this is no problem even without such a flag if you have a fast swap left/right view option by hotkey like in my plugin. So I'll make and test how it works.

**Comment by VRS3DGuru2** [comment]
I agree, it is easy to just swap eyes with a hotkey.  The display wont receive a message to activate 3D mode on the display from this approach, but that is also ok, there is a button the remotes to activate it on the display once the game starts.  I think those 2 situations are completely ok to accept. I tried adding an independent sequential method to your plugin and couldn't figure it out. I am sure I am missing something in how to swap cameras between frames, I think that VSYNC will also have to be turned on to make sure that player settings don't break it.
This example from Unity works for VSYNC and attempt to lock it to 120Hz -
`void Start()
    {
        // Sync the frame rate to the screen's refresh rate
        QualitySettings.vSyncCount= 1;
       // Make the game run as fast as possible
        Application.targetFrameRate = 1;
        // Limit the framerate to 60
        Application.targetFrameRate= 120;
    }`

**Comment by Vital-Volkov** [comment]
Just add a bool like leftView=true/false and reverse it in Update. Then add a sequential method to the shader and supply the bool as leftView=1/0 and make the shader show a full left/right image based on the 1/0 switcher.

Yes, vSync is required for sequential and for Over/Under.

**Comment by Xeyos88** [comment]
> Just add a bool like leftView=true/false and invert it in Update. Then add a sequential method to the shader and supply the bool as leftView=1/0 and make the shader show a full left/right image based on the 1/0 switcher.
> 
> Yes, vSync is required for sequential and for Over/Under.

Hi, do have an example code of this solution?

**Comment by Vital-Volkov** [comment]
> Hi, do have an example code of this solution?

Hi, no but I'll make it when I have active S3D hardware. 
I already finished the new Unity S3D version and am testing it now on different demos before release.

**Comment by SinanAkkoyun** [comment]
I am also very interested in using sequential rendering

How exactly would we need to modify the shader?

**Comment by Vital-Volkov** [comment]
OK I think add this for testing in coming new version

**Comment by Vital-Volkov** [comment]
Hi!
Added Method.Sequential test. 
Should work with shutter glasses if VSync is on and FPS equal to refresh rate of monitor.

**Comment by BrunoItaly** [comment]
Hello,
I have a 3d projection system with 120hz projector, nvidia Quadro Gpu, Volfoni Glasses and ART tracking system, which I use for VR presentations (non headset) and I tryed to implement your amazing plugin on Unity 2022 but had no luck so far.

I Downloaded the Stereo3D&Direct3D11.1_Unity_Demo and I can run it  in 3D without any problem on my system (rendering Pc has W11) but I was not able to create  myself a build on Unity 2022.3 using the "Stereoscopic-3D-system-for-Unity-2019--Direct3D11.1" distribution or even the main branch. 
Tried with just one camera and with two cameras and in the build I see only the GUI window but i get only a mono rendering.
For sure I'm doing something wrong or forgot some requirements. It should be great if you could share more info on how to implement your S3d in a Unity project (maybe sharing the demo project will be the easiest).

**Comment by Vital-Volkov** [comment]
Hi,
I am now making native S3D plugin and also found black screen with "GUI as overlay" and other script errors while testing on Unity 2019, 2020 so I'll solve this problem.
I made [native D3D11 Stereo3D](https://github.com/Vital-Volkov/Remake-of-Direct3D11-native-stereoscopic-sample) working with Unity directly via C++ plugin so no more need "stereo display non head-mounted" support, anyway builtin "separation/convergence S3D" settings approach is incorrect and mess with "single/multi pass" rendering not working as it should.
But now all is clean and easy, Unity only makes render to 2 textures and sends its pointers to the plugin which do all other work by loading textures into back buffers of D3D11 or OpenGL so I think I'll add "quad buffer" support also.

**Comment by BrunoItaly** [comment]
Hello,
It’s an important job IMHO,  I saw your native D3D11 Stereo3D<https://urlsand.esvalabs.com/?u=https%3A%2F%2Fgithub.com%2FVital-Volkov%2FRemake-of-Direct3D11-native-stereoscopic-sample&e=4641458d&h=a26ff5ea&f=n&p=y>  but I think that Unity plugin and quad buffer is still work in progress.
I will look at your github page,

PS. do not understand why Unity is not developing this part internally but if you create something working, I think it will be useful for many customers.


Bruno Arina
More srl
Sede legale
Strada della Campanara, 15
61122 Pesaro (PU)

Sede operativa
Via L. Manara, 136
63074 S. Benedetto del Tronto (AP)
Tel.  0736 618160
Mobile  +39 331 5092015
Email:   ***@***.******@***.***>
Web:    www.more.mr<http://www.more.mr/>

Non stampare questa e-mail.

Questo documento è formato esclusivamente per il destinatario. Tutte le informazioni ivi contenute, compresi eventuali allegati, sono da ritenere esclusivamente confidenziali e riservate secondo i termini del vigente D.Lgs. 196/2003 in materia di privacy e del Regolamento europeo 679/2016 – GDPR-  e quindi ne è proibita l’utilizzazione ulteriore non autorizzata. Se avete ricevuto per errore questo messaggio, Vi preghiamo cortesemente di contattare immediatamente il mittente e cancellare la e-mail. Grazie.

Please don’t print this e-mail.

Confidentiality Notice – This e-mail message including any attachments is for the sole use of the intended recipient and may contain confidential and privileged information pursuant to Legislative Decree 196/2003 and the European General Data Protection Regulation 679/2016 – GDPR-. Any unauthorized review, use, disclosure or distribution is prohibited. If you are not the intended recipient, please contact the sender by reply e-mail and destroy all copies of the original message.

Da: Vital ***@***.***>
Inviato: domenica 24 marzo 2024 11:26
A: Vital-Volkov/Stereoscopic-3D-system-for-Unity-2019- ***@***.***>
Cc: Bruno Arina ***@***.***>; Comment ***@***.***>
Oggetto: Re: [Vital-Volkov/Stereoscopic-3D-system-for-Unity-2019-] Unity 2022.2+ DX11/12 3D Active Stereo (Issue #7)

Questa è la prima volta che ricevi un'email da questo mittente. Assicurati che sia qualcuno di cui ti fidi.

Hi,
I am now making native S3D plugin and also found black screen with "GUI as overlay" and other script errors while testing on Unity 2019, 2020 so I'll solve this problem.
I made native D3D11 Stereo3D<https://urlsand.esvalabs.com/?u=https%3A%2F%2Fgithub.com%2FVital-Volkov%2FRemake-of-Direct3D11-native-stereoscopic-sample&e=4641458d&h=a26ff5ea&f=n&p=y> working with Unity directly via C++ plugin so no more need "stereo display non head-mounted" support, anyway builtin "separation/convergence S3D" settings approach is incorrect and mess with "single/multi pass" rendering not working as it should.
But now all is clean and easy, Unity only makes render to 2 textures and sends its pointers to the plugin which do all other work by loading textures into back buffers of D3D11 or OpenGL so I think I'll add "quad buffer" support also.

—
Reply to this email directly, view it on GitHub<https://urlsand.esvalabs.com/?u=https%3A%2F%2Fgithub.com%2FVital-Volkov%2FStereoscopic-3D-system-for-Unity-2019-%2Fissues%2F7%23issuecomment-2016762865&e=4641458d&h=46de38b6&f=n&p=y>, or unsubscribe<https://urlsand.esvalabs.com/?u=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FBHG7CTUTGFZOCDZO7CY4IL3YZ2S3VAVCNFSM6AAAAAAWWY3W2WVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMJWG43DEOBWGU&e=4641458d&h=f1cc5356&f=n&p=y>.
You are receiving this because you commented.Message ID: ***@***.******@***.***>>

--
Questo messaggio e' stato analizzato con Libraesva ESG ed e' risultato non infetto.
Clicca qui per segnalarlo come spam.<https://mail.websolute.com/action/4V2XK74lBfz5yPL/learn-spam>
Clicca qui per metterlo in blocklist<https://mail.websolute.com/action/4V2XK74lBfz5yPL/blocklist>

**Comment by Vital-Volkov** [comment]
My answer to mail may be useful for others:

```
I'll add a more detailed manual later.

PPI is the hardware spec parameter of each screen and it's constant and required for all other settings precision calculated in real millimeters 
so if you once input real values of your screen PPI and your eyes IPD then you get 100% depth same as real life S3D and never overshoot your eyes farther then parallel optic axes to outside(negative) angle 
which easily happens with unscaled common settings where you need to check unknown settings IPD result with a ruler on screen on each game or viewport size change.
My system does all of this automatically based on calculated viewport size(as it knows real screen size based on PPI and native resolution of screen) and no matter what texture resolution static or dynamic 
all calculations do in millimeters. So PPI is critical for precision and you should know the PPI of your screen. Make correct settings once and save it for each user and you get 100% S3D depth like in real life 
and then play only with Virtual IPD which only affect world scale S3D perception how it see small or big creatures(or human with special optic) with different stereo base in real life and no matter what virtual 
creature you emulate max S3D depth will be same 100% if you set you real correct PPI and user IPD, period. PPI is critical and should always be real to match all other sizes to real millimeters. But then 
you sure can set the User IPD lower than your real IPD and get not a 100% depth pseudo S3D not like in real life(you will see star sky like in room wall depth instead of infinity). 
The Main thing is NEVER overshoot your IPD.
So I always using real PPI and User IPD settings as constants and changing only Virtual IPD for fun or just leave Virtual IPD match to User IPD for real view and 
if you also adjust FOV until Screen Distance match your real distance from eyes to screen then you get precision match virtual world with real. 
But for HMD Screen Distance is not real due to lenses magnification so I think calculations should be like these: Real HMD screen distance from eyes divide to 
lenses magnification for example 60mm / 3x = 20mm or just set FOV from HMD specs if it correct.
```

**Comment by mfleurent** [comment]
> Hi, I am now making native S3D plugin and also found black screen with "GUI as overlay" and other script errors while testing on Unity 2019, 2020 so I'll solve this problem. I made [native D3D11 Stereo3D](https://github.com/Vital-Volkov/Remake-of-Direct3D11-native-stereoscopic-sample) working with Unity directly via C++ plugin so no more need "stereo display non head-mounted" support, anyway builtin "separation/convergence S3D" settings approach is incorrect and mess with "single/multi pass" rendering not working as it should. But now all is clean and easy, Unity only makes render to 2 textures and sends its pointers to the plugin which do all other work by loading textures into back buffers of D3D11 or OpenGL so I think I'll add "quad buffer" support also.

Hello there,

first of, congrats for the huge work on stereo3D, it's really fascinating.
I see you mention Unity native plugin to produce true quad buffer, it's exactly what i try to do.... 

I have an old DLP projector with active shutter glass from a decommissioned powerwall, i try to make it work in sequential mode.
it kind of work with your Stereo3D script but the glass sync is very fragile and become weird very quickly. I think te way to go is quad buffer trough a unity native plugin, preferably on openGL because the quadro GPU is on a Win7 OS.

I began to work on that but i cant manage to use the texture pointer from unity to inject it into th openGL window.

it seem's like you already done this job so, any advices or example tat i can learn from is welcome.

Thanks
Martin

**Comment by Vital-Volkov** [comment]
Hi [mfleurent](https://github.com/mfleurent), thanks!

Yes, `Sequential` output in my S3D system should work fine only if FPS is stable over monitor refresh and VSync is On.
I am making a D3D11 native C++ plugin now for my S3D system but I plan to also make it for OpenGL.
I read about a problem with Unity pixel format set on window creation for OpenGL quad buffer S3D and code injection via hook required, do you already solve this?

I found that you can directly use Unity render texture in native C++ if the format is `RGB111110Float` and you set `RenderTexture.Create()` before using the pointer. If I use common render texture format then I have black screen and get render texture to show in C++ only after making the GPU copy texture from Unity render texture first. It drops FPS a little but it works also to show this copied texture as a shared resource with a second D3D device.
